lib/auth/auth.go

```diff
@@ -1666,6 +1698,15 @@ func certRequestDeviceExtensions(ext tlsca.DeviceExtensions) certRequestOption {
 	}
 }
 
+// getUserOrLoginState will return the given user or the login state associated with the user.
+func (a *Server) getUserOrLoginState(ctx context.Context, user types.User) services.UserState {
+	if uls, _ := a.GetUserLoginState(ctx, user.GetName()); uls != nil {
+		return uls
+	}
+
+	return user
+}
```

Comment:
IMO we shouldn't be loading the user resource at all for access-control logic unless we explicitly get a not found error, it will be too easy to accidentally use the pre-mapping state in cases where it would be wrong to do so (which are most cases now... user login state effectively replaces `types.User` for almost everything in `auth_with_roles.go`).

Off the top of my head, I'm fairly sure only cases where we should ever have to load the legacy user state for access-control decisions should be either:

1. We just upgraded and there are users still logged in from before the login state feature rolled out.
2. The identity in question is a bot or other piece of automation whose certs were created directly rather than being the result of a normal "login" process.

Based on that, I'd suggest reworking this function to something like this:

```golang
// getUserOrLoginState will return login state for a user if available, falling back to returning the user resource.
func (a *Server) getUserOrLoginState(ctx context.Context, username string) (services.UserState, error) {
	uls, err := a.GetUserLoginState(ctx, username)
	if err != nil && !trace.IsNotFound(err) {
		return nil, trace.Wrap(err)
	}
	
	if err == nil {
		return uls, nil
	}

	user, err := a.GetUser(username,  false)
	return user, trace.Wrap(err)
}
```

Reply:
I've implemented this change and a number of the downstream impacts here. I'll resolve this, but I wouldn't be surprised if there were some additional issues.